third_party/terraform/data_sources/data_source_google_iam_testable_permissions.go

```diff
@@ -0,0 +1,157 @@
+		Schema: map[string]*schema.Schema{
+			"full_resource_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"custom_support_level": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "SUPPORTED",
+			},
+			"stage": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "GA",
+			},
+			"permissions": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"title": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"custom_support_level": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"stage": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"api_disabled": {
+							Type:     schema.TypeBool,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceGoogleIamTestablePermissionsRead(d *schema.ResourceData, meta interface{}) (err error) {
+	config := meta.(*Config)
+	body := make(map[string]interface{})
+	body["pageSize"] = 500
+	permissions := make([]map[string]interface{}, 0)
+	custom_support_level := d.Get("custom_support_level").(string)
+	if err = validatePermissionCustomSupport(custom_support_level); err != nil {
+		return err
+	}
+
+	stage := d.Get("stage").(string)
+	if err = validatePermissionStage(stage); err != nil {
+		return err
+	}
+
+	for {
+		url := "https://iam.googleapis.com/v1/permissions:queryTestablePermissions"
+		body["fullResourceName"] = d.Get("full_resource_name").(string)
+		res, err := sendRequest(config, "POST", "", url, body)
+		if err != nil {
+			return fmt.Errorf("Error retrieving permissions: %s", err)
+		}
+
+		pagePermissions := flattenTestablePermissionsList(res["permissions"], custom_support_level, stage)
+		permissions = append(permissions, pagePermissions...)
+		pToken, ok := res["nextPageToken"]
+		if ok && pToken != nil && pToken.(string) != "" {
+			body["pageToken"] = pToken.(string)
+		} else {
+			break
+		}
+	}
+
+	if err := d.Set("permissions", permissions); err != nil {
+		return fmt.Errorf("Error retrieving permissions: %s", err)
+	}
+
+	d.SetId(d.Get("full_resource_name").(string))
+	return nil
+}
+
+func flattenTestablePermissionsList(v interface{}, custom_support_level string, stage string) []map[string]interface{} {
+	if v == nil {
+		return make([]map[string]interface{}, 0)
+	}
+
+	ls := v.([]interface{})
+	permissions := make([]map[string]interface{}, 0, len(ls))
+	for _, raw := range ls {
+		p := raw.(map[string]interface{})
+
+		if _, ok := p["name"]; ok {
+			var csl bool
+			if custom_support_level == "SUPPORTED" {
+				csl = p["customRolesSupportLevel"] == nil || p["customRolesSupportLevel"] == "SUPPORTED"
+			} else {
+				csl = p["customRolesSupportLevel"] == custom_support_level
+			}
+
+			if csl && p["stage"] == stage {
```

Comment:
Is there any concept of a stage hierarchy here? For instance a "GA" permission could be available in "BETA"?

Reply:
Good question....and I thought about that but I think I like the explicit nature of having it be a filter for BETA as opposed to assuming users want GA and BETA since it's still possible to instantiate 2 data providers (GA and BETA), but it might be a bit more difficult to further filter if we assume `BETA == GA || BETA`. What are your thoughts?

Reply:
I think it depends on the behavior of the API here. In my experience with other APIs and developing the beta/GA providers beta is almost always a superset of GA, so anything in GA is also in beta. 

Is it implied that you can use a GA role in a beta context? For example does the response look like this:
GA roles for some resource:
`roles/owner`,
`roles/editor`

Beta roles:
`roles/newRole`

Or would beta roles include all GA roles like this:
Beta roles:
`roles/owner`,
`roles/editor`,
`roles/newRole`

Reply:
The behavior of the API is that BETA is not a superset of GA. Once a permission is GA, the enum changes to GA and it's no longer in BETA

Reply:
Fixed by making `stages` a slice that can take multiple stage identifiers